src/User/Controller/SettingsController.php

```diff
@@ -435,20 +440,31 @@ public function actionDelete()
 
     public function actionTwoFactor($id)
     {
+        $choice=Yii::$app->request->post('choice');
         /**
-        * 
-        *
         * @var User $user 
-        */
+        */        
         $user = $this->userQuery->whereId($id)->one();
 
         if (null === $user) {
             throw new NotFoundHttpException();
         }
-
-        $uri = $this->make(TwoFactorQrCodeUriGeneratorService::class, [$user])->run();
-
-        return $this->renderAjax('two-factor', ['id' => $id, 'uri' => $uri]);
+        
+        switch($choice)
+        {
+            case 'google-authenticator':
+                $uri = $this->make(TwoFactorQrCodeUriGeneratorService::class, [$user])->run();
+                return $this->renderAjax('two-factor', ['id' => $id, 'uri' => $uri]);
+            case 'email':
+                $emailCode = $this->make(TwoFactorEmailCodeGeneratorService::class, [$user])->run();
+                return $this->renderAjax('two-factor-email', ['id' => $id, 'code' => $emailCode]);
+            case 'sms':
+                // get mobile phone, if exists
+                $mobilePhone=$user->getAuthTfMobilePhone();
+                $smsCode = $this->make(TwoFactorSmsCodeGeneratorService::class, [$user])->run();
```

Comment:
Why are you generating the code here and in `actionTwoFactorMobilePhone` both? This way it ties to send the text message two times and the first time, since it doesn't have a number to send it to yet, it always displays an error message when reloading the page.